blockindex/sgd.go

```diff
@@ -0,0 +1,235 @@
+// Copyright (c) 2022 IoTeX Foundation
+// This source code is provided 'as is' and no warranties are given as to title or non-infringement, merchantability
+// or fitness for purpose and, to the extent permitted by law, all liability for your use of the code is disclaimed.
+// This source code is governed by Apache License 2.0 that can be found in the LICENSE file.
+
+package blockindex
+
+import (
+	"context"
+	"errors"
+	"time"
+
+	"github.com/iotexproject/go-pkgs/cache"
+	"github.com/iotexproject/go-pkgs/hash"
+	"github.com/iotexproject/iotex-address/address"
+	"github.com/iotexproject/iotex-proto/golang/iotextypes"
+	"google.golang.org/protobuf/proto"
+
+	"github.com/iotexproject/iotex-core/action"
+	"github.com/iotexproject/iotex-core/blockchain/block"
+	"github.com/iotexproject/iotex-core/blockchain/blockdao"
+	"github.com/iotexproject/iotex-core/blockindex/indexpb"
+	"github.com/iotexproject/iotex-core/db"
+	"github.com/iotexproject/iotex-core/db/batch"
+	"github.com/iotexproject/iotex-core/pkg/util/byteutil"
+)
+
+const (
+	_sgdBucket                = "sgd"
+	_sgdToHeightNS            = "hh"
+	_sgdActCreate  sgdActType = iota
+	_sgdActCall
+)
+
+var _sgdCurrentHeight = []byte("currentHeight")
+
+type (
+	// SGDIndexer is the interface for Sharing of Gas-fee with DApps indexer
+	SGDIndexer interface {
+		blockdao.BlockIndexer
+		// CheckContract
+		CheckContract(contract string) (address.Address, uint64, bool)
+		// GetSGDIndex
+		GetSGDIndex(contract string) (*indexpb.SGDIndex, error)
+	}
+
+	sgdRegistry struct {
+		kvStore    db.KVStore
+		kvCache    cache.LRUCache
+		percentage uint64
+	}
+
+	sgdActType int
+
+	sgdAct struct {
+		actType    sgdActType
+		sender     address.Address
+		contract   string
+		createTime time.Time
+	}
+)
+
+// NewSGDRegistry creates a new SGDIndexer
+func NewSGDRegistry(kv db.KVStore, kvCache cache.LRUCache, percentage uint64) SGDIndexer {
```

Comment:
as commented above, remove precentage
also remove `kvCache`, it can be generated inside here
you can use a builder config to pass in the parameters needed, refer to https://github.com/iotexproject/iotex-core/commit/02f8190fa7125e032e59725bb2b594a4351e8949#diff-c65dd5a657c73eb89458c69a82a25fc0993655650b3ad477136955c31e50e2bb

Reply:
if inside generate `kvCache`, The test cannot cover the situation without cache.